samples/sample_common/include/sample_utils.h

```diff
@@ -625,6 +696,207 @@ void init_ext_buffer(T & ext_buffer)
+    {
+        ExtBufIterator it = std::find_if(m_ext_buf.begin(), m_ext_buf.end(), CmpExtBufById(id));
+        if (it != m_ext_buf.end())
+        {
+            delete [] (mfxU8*)(*it);
+            m_ext_buf.erase(it);
+            RefreshBuffers();
+        }
+    }
+
+    template <typename TB>
+    TB* GetExtBuffer()
+    {
+        ExtBufIterator it = std::find_if(m_ext_buf.begin(), m_ext_buf.end(), CmpExtBufById(mfx_ext_buffer_id<TB>::id));
+        if (it != m_ext_buf.end())
+        {
+            return (TB*)*it;
+        }
+        return 0;
+    }
+
+    template <typename TB>
+    operator TB*()
+    {
+        return (TB*)GetExtBuffer(mfx_ext_buffer_id<TB>::id);
+    }
+
+    template <typename TB>
+    operator TB*() const
+    {
+        return (TB*)GetExtBuffer(mfx_ext_buffer_id<TB>::id);
+    }
+
+private:
+    mfxExtBuffer* AddExtBuffer(mfxU32 id, mfxU32 size)
+    {
+        if(!size || !id)
+            return 0;
+
+        // Limitation: only one ExtBuffer instance can be stored
+        ExtBufIterator it = std::find_if(m_ext_buf.begin(), m_ext_buf.end(), CmpExtBufById(id));
+        if (it == m_ext_buf.end())
+        {
+            mfxExtBuffer* buf = (mfxExtBuffer*)new mfxU8[size];
+            memset(buf, 0, size);
+            m_ext_buf.push_back(buf);
+
+            mfxExtBuffer& ext_buf = *buf;
+            ext_buf.BufferId = id;
+            ext_buf.BufferSz = size;
+            RefreshBuffers();
+
+            return m_ext_buf.back();
+        }
+
+        return *it;
+    }
+
+    mfxExtBuffer* GetExtBuffer(mfxU32 id) const
+    {
+        constExtBufIterator it = std::find_if(m_ext_buf.begin(), m_ext_buf.end(), CmpExtBufById(id));
+        if (it != m_ext_buf.end())
+        {
+            return *it;
+        }
+        return 0;
+    }
+
+    void RefreshBuffers()
+    {
+        this->NumExtParam = (mfxU32)m_ext_buf.size();
+        this->ExtParam    = this->NumExtParam ? m_ext_buf.data() : 0;
+    }
+
+    void ClearBuffers()
+    {
+        if (0 != m_ext_buf.size())
+        {
+            for (ExtBufIterator it = m_ext_buf.begin(); it != m_ext_buf.end(); it++ )
+            {
+                delete [] (mfxU8*)(*it);
+            }
+            m_ext_buf.clear();
+        }
+        RefreshBuffers();
+    }
+
+    bool IsCopyAllowed(mfxU32 buf_id)
+    {
+        static mfxU32 allowed[] = {
+            MFX_EXTBUFF_CODING_OPTION,
+            MFX_EXTBUFF_CODING_OPTION2,
+            MFX_EXTBUFF_CODING_OPTION3,
+            MFX_EXTBUFF_FEI_PARAM,
+            MFX_EXTBUFF_BRC,
+            MFX_EXTBUFF_HEVC_PARAM,
+        };
+
+        for (mfxU32 i = 0; i < sizeof(allowed)/sizeof(allowed[0]); i++)
+        {
+            if (buf_id == allowed[i])
+                return true;
+        }
+
+        return false;
+    }
+
+    typedef std::vector<mfxExtBuffer*>::iterator ExtBufIterator;
+    typedef std::vector<mfxExtBuffer*>::const_iterator constExtBufIterator;
+    std::vector<mfxExtBuffer*> m_ext_buf;
```

Comment:
I think we should avoid all low-level memory operations as much as possible, so we could have some other member for storage:
```
std::vector<std::unique_ptr<mfxU8[]>> m_memory_storage;
```